base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -785,15 +786,99 @@ public void validateRevocation(ACMEAccount account, ACMERevocation revocation) t
             // No need to check order status since it's guaranteed to be valid.
             // No need to check order expiration since it's irrelevant for revocation.
 
-            logger.info("Account authorized to revoke certificate");
+            logger.info("Account issued the certificate");
             return;
         }
 
         logger.info("Order not found");
 
-        // TODO: Case 2: validate using cert identifiers authorizations
+        // TODO: Case 2: validate using authorization records (if they still exist)
+
+        logger.info("Getting certificate identifiers");
+        Collection<ACMEIdentifier> identifiers = getCertIdentifiers(cert);
+
+        logger.info("Getting account authorizations");
+        Date now = new Date();
+
+        for (ACMEAuthorization authz : database.getAuthorizationsByAccount(account.getID())) {
+
+            String status = authz.getStatus();
+            if (!"valid".equals(status)) {
+                logger.info("Authorization " + authz.getID() + " is " + status);
+                continue;
+            }
+
+            Date expirationTime = authz.getExpirationTime();
+            if (!expirationTime.after(now)) {
+                logger.info("Authorization " + authz.getID() + " has expired");
+                continue;
+            }
+
+            ACMEIdentifier identifier = authz.getIdentifier();
+            logger.info("Authorization " + authz.getID() + " is valid for " + identifier);
+
+            // remove authorized identifier from the list
+            // TODO: handle wildcard
+
+            identifiers.remove(identifier);
+        }
+
+        if (identifiers.isEmpty()) {
+            logger.info("Account holds the required authorizations");
+            return;
+        }
+
+        logger.info("Missing authorizations for:");
+        for (ACMEIdentifier identifier : identifiers) {
+            logger.info("- " + identifier.getType() + ": " + identifier.getValue());
+        }
 
         // TODO: generate proper exception
         throw new Exception("Account unauthorized to revoke certificate");
     }
+
+    public Collection<ACMEIdentifier> getCertIdentifiers(X509CertImpl cert) throws Exception {
+
+        // use HashSet to store cert identifiers without duplicates
+        Collection<ACMEIdentifier> identifiers = new HashSet<>();
+
+        X500Name subjectDN = cert.getSubjectObj().getX500Name();
+        logger.info("Subject DN: " + subjectDN);
+
+        String cn;
+        try {
+            cn = subjectDN.getCommonName();
+
+        } catch (NullPointerException e) {
+            // X500Name.getCommonName() throws NPE if subject DN is blank
+            // TODO: fix X500Name.getCommonName() to return null
```

Comment:
Off topic, but it would better to have `X500Name.getCommonName()` return a `List<String>`, which covers the case of zero, one, or multiple (ordered) CommonName attributes in the DN.  In the "empty subject DN" case, the empty list would be returned.

Enough with `null`; if we change it we might as well move to a more correct model of X.509 with a more robust implementation :)

Reply:
I thought X.500 only allows a single common name, but regardless, PKI will use whatever interface provided by JSS.

Reply:
Oh wow, this is in the old `netscape.security` code. 

And TBF @frasertweedale, that's the spec of `X500Name.getCommonName`:

```java
    /**
     * Returns a "Common Name" component. If more than one such
     * attribute exists, the topmost one is returned.
     *
     * @return "CN=" component of the name, if any.
     */
    public String getCommonName() throws IOException {
        DerValue attr = findAttribute(commonName_oid);

        return getString(attr);
    }
```

So if you wanted to do what you were suggesting (multiple `CommonName`s), you'd instead do:

```
subjectDN.getAttributesForOid(X500Name.commonName_oid)
```

I think. :-) Code not tested.

Reply:
Fair enough.  I like precise APIs and prefer decisions like "give me the first one" be pushed out to the library user.  But at least the behaviour is documented!